icon_validator/validate.py

```diff
@@ -11,47 +11,56 @@
 from .timing import *
 
 
-# TODO refactor this into a class to allow for easier and better testing
-def validate(directory, spec_file_name="plugin.spec.yaml", fail_fast=False, run_all=False, validators=list()):
-    spec = KomandPluginSpec(directory, spec_file_name)
-    status = 0  # Resultant return code
-    start_time = time_now()
-    print(f"{BULLET_OK} {BOLD}Running Integration Validators...{CEND}")
-
-    if not validators:
-        if spec_file_name == "plugin.spec.yaml":
-            validators = VALIDATORS
-            if run_all:
-                validators += JENKINS_VALIDATORS
-        elif spec_file_name == "workflow.spec.yaml":
-            validators = WORKFLOW_VALIDATORS
-
-    for v in validators:
-        print(f"{BULLET_OK} Executing validator {v.name}")
-        try:
-            v.validate(spec)
-            success = True
-
-        # TODO refactor to use ValidationException will require refactoring of all plugin validators first
-        except Exception as e:
-            print(f"Validator {v.name} failed!")
-            ex_type, ex, tb = sys.exc_info()
-            traceback.print_exception(Exception, e, tb)
-            status = 1
-            success = False
-
-        if not success and fail_fast:
-            break
-
-    end_time = time_now()
-    time_elapsed = format_time(start=start_time, end=end_time)
-
-    extension = spec_file_name.split(".")[0].capitalize()
-
-    if status == 0:
-        print(f"{BULLET_OK} {BOLD}{extension} successfully validated!{CEND}")
-    else:
-        print(f"{BULLET_FAIL}{extension} failed validation!")
-
-    print(f"\n----\n{BULLET_OK}{BOLD} Total time elapsed: {time_elapsed}ms{CEND}")
-    return status
+class Validate:
+
+    def __init__(self, validators=None, unit_test=False):
+        if validators is None:
+            validators = list()
+        self.validators = validators
+        self.unit_test = unit_test  # Used to stop some of the print outs during unit testing, to increase legibility
+
+    def validate(self, directory, spec_file_name="plugin.spec.yaml", fail_fast=False, run_all=False):
```

Comment:
Not much is gained from using a class to be worth breaking backwards compatibility and updating our other tools. I am going to say we should maintain backwards compatibility until we can no longer do so.

What you can do instead is update the validate function declaration to the following:
```
def validate(self, directory, spec_file_name="plugin.spec.yaml", fail_fast=False, run_all=False, validators=[], unit_test=False):
```

OR

```
def validate(self, directory, spec_file_name="plugin.spec.yaml", fail_fast=False, run_all=False, validators=[]):
```
And then within the function, check for an environment variable that will be set by the unit test `setUp` function, something like `os.environ["ICON_VALIDATOR_IS_UNIT_TEST"]`

That's my $0.02. The benefit here is not worth the time spent updating existing tools using the validators.

Reply:
I am torn on this....I like the work done here I think it is cleaner and I am a sucker for a class.  But I think Mike has a strong point on this.  I think we should consider backwards compatibility when releasing products that will be used by the public.

Reply:
I approved this but I think maybe a group chat on this maybe important.  We are having talks in the slack chat about improving accessibility.  If there is code out there that this will break, I think that hurts our image in the kommunity.